kani-compiler/src/codegen_cprover_gotoc/codegen/contract.rs

```diff
@@ -0,0 +1,87 @@
+use crate::codegen_cprover_gotoc::GotocCtx;
+use cbmc::goto_program::{Contract, Expr, Spec, Symbol};
+use rustc_ast::NestedMetaItem;
+use rustc_middle::mir::Local;
+use rustc_middle::ty::{self, FnSig};
+
+impl<'tcx> GotocCtx<'tcx> {
+    /// Transforms current function's arguments into expressions
+    fn extract_function_arguments(&mut self, sig: FnSig<'tcx>) -> Vec<Expr> {
+        let args = sig.inputs();
+        args.iter()
+            .enumerate()
+            .map(|(i, t)| {
+                let lc = Local::from_usize(i + 1); // The zeroth index (local variable) is reserved for the return value of the function. Hence, function argument index starts from one.
+                let ident = self.codegen_var_name(&lc);
+                let typ = self.codegen_ty(*t);
+                Expr::symbol_expression(ident, typ)
+            })
+            .collect()
+    }
+
+    /// Transforms current function's return value into an expression
+    fn extract_function_return_value(&mut self, sig: FnSig<'tcx>) -> Expr {
+        let rt = sig.output();
+        let lc = Local::from_usize(0); // The zeroth index (local variable) is reserved for the return value of the function.
+        let ident = self.codegen_var_name(&lc);
+        let typ = self.codegen_ty(rt);
+        Expr::symbol_expression(ident, typ)
+    }
+
+    /// In order to keep the contract symbol self-contained, every function contract clause
+    /// such as `requires`, `ensures`, etc. is wrapped into a lambda expression.
+    /// The binding variables of these lambda expressions are -
+    ///  1) the return value of the function as the first variable, followed by,
+    ///  2) the list of function arguments (in order).
+    fn get_lambda_binding_variables(&mut self, sig: FnSig<'tcx>) -> Vec<Expr> {
+        let mut vars = vec![];
+        let ret = self.extract_function_return_value(sig);
+        vars.push(ret);
+        let args = self.extract_function_arguments(sig);
+        vars.extend(args);
+        return vars;
+    }
+
+    /// Generates a specification containing a lambda expression with "true" as its body
+    fn spec_true(&mut self, sig: FnSig<'tcx>) -> Spec {
+        let bv = self.get_lambda_binding_variables(sig);
+        let mir = self.current_fn().mir();
+        let loc = self.codegen_span(&mir.span);
+        Spec::new(bv, Expr::bool_true(), loc)
+    }
+
+    /// Generates a symbol for the function contract and adds it to the symbol table
+    pub fn codegen_assigns_clause(&mut self, args: Vec<NestedMetaItem>) {
+        let mir = self.current_fn().mir();
+        let loc = self.codegen_span(&mir.span);
+        let sig = self.current_fn().sig();
+        let sig =
+            self.tcx.normalize_erasing_late_bound_regions(ty::ParamEnv::reveal_all(), sig.unwrap());
+        let typ = self.fn_typ();
+
+        let name = format!("contract::{}", self.current_fn().name()); // name of the contract symbol
```

Comment:
Can a function be annotated with multiple assigns annotations? If it can, you should probably consider adding an id here. Otherwise, please check that's not the case and emit an error if the user tries to use 2 or more annotations.

Reply:
Yes, a function can only have one contract. Each contract can have multiple `requires`, `ensures` and `assigns` annotations. They will all be treated as one contract specification collectively. 

If multiple `requires` or `ensures` statements are present, then they will get inlined as separate `assert` or `assume` statements during macro expansion (will send in separate PR). 

Multiple `#[kani::assigns]` statements get wrapped into one `#[kanitool::assigns(..)]` during macro expansion. Hence, we can assume that there is one assigns clause here.